### PR TITLE
Fast-SCNN implemented

### PR DESCRIPTION
**Pull request by johnzja**
69.1% mIoU achieved, compared with 68% according to the original paper. Learning rate of 0.12 is suggested after multiple experimental trainings. However, three failures occurs when running pytest (failures are irrelevant with fast-scnn itself).

### CODE REVIEW

File: mmseg/ops/separable_conv_module.py [hunk 1]
```diff
@@ -1,88 +1,60 @@
-import torch.nn as nn
-from mmcv.cnn import ConvModule
+from mmcv.cnn import build_norm_layer
+from torch import nn
 
 
 class DepthwiseSeparableConvModule(nn.Module):
```
Comment (xvjiarui) [inline mmseg/ops/separable_conv_module.py:5]: We may keep this file unchanged.

File: mmseg/models/decode_heads/sep_fcn_head.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+from mmseg.ops import DepthwiseSeparableConvModule
+from ..builder import HEADS
+from .fcn_head import FCNHead
+
+
+@HEADS.register_module()
+class SepFCNHead(FCNHead):
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/sep_fcn_head.py:7]: Rename to `DepthwiseSeparableFCNHead`.

File: mmseg/utils/collect_env.py [hunk 3]
```diff
@@ -41,9 +42,12 @@ def collect_env():
...
 
-    gcc = subprocess.check_output('gcc --version | head -n1', shell=True)
-    gcc = gcc.decode('utf-8').strip()
-    env_info['GCC'] = gcc
+    try:
+        gcc = subprocess.check_output('gcc --version | head -n1', shell=True)
+        gcc = gcc.decode('utf-8').strip()
+        env_info['GCC'] = gcc
+    except subprocess.CalledProcessError:
+        env_info['GCC'] = 'n/a'
```
Comment (xvjiarui) [inline mmseg/utils/collect_env.py:50]: Are these lines necessary?

File: mmseg/models/backbones/fast_scnn.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        x = self.bottleneck1(x)
+        x = self.bottleneck2(x)
+        x = self.bottleneck3(x)
+        x = torch.cat([x, *self.ppm(x)], dim=1)
+        x = self.out(x)
+        return x
+
+
+class FeatureFusionModule(nn.Module):
+    """Feature fusion module."""
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:118]: We may use standard doc string here.

File: mmseg/models/backbones/fast_scnn.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+                 higher_in_channels=64,
+                 lower_in_channels=128,
+                 fusion_out_channels=128,
+                 scale_factor=4,
+                 out_indices=(0, 1, 2),
+                 conv_cfg=None,
+                 norm_cfg=dict(type='BN'),
+                 act_cfg=dict(type='ReLU'),
+                 align_corners=False):
+        """Fast-SCNN Backbone.
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:191]: Move docstring above `__init__()`

File: mmseg/models/backbones/mobile_net_v2.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+    def forward(self, x):
+        if self.use_res_connect:
+            return x + self.conv(x)
+        else:
+            return self.conv(x)
+
+
+@BACKBONES.register_module()
+class MobileNetV2(nn.Module):
```
Comment (xvjiarui) [inline mmseg/models/backbones/mobile_net_v2.py:67]: `MobileNetV2` is not supported yet. We may delete this file for now.

File: configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
+_base_ = [
+    '../_base_/models/fast_scnn.py', '../_base_/datasets/cityscapes.py',
+    '../_base_/default_runtime.py'
+]
+crop_size = (512, 1024)
+cudnn_benchmark = True
+# model training and testing settings
+train_cfg = dict()
+test_cfg = dict(mode='whole')
```
Comment (xvjiarui) [inline configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py:9]: Move these to `configs/_base_/models/fast_scnn.py`.

File: configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            dict(type='ImageToTensor', keys=['img']),
+            dict(type='Collect', keys=['img']),
+        ])
+]
+data = dict(
+    samples_per_gpu=8,
+    workers_per_gpu=4,
+    train=dict(pipeline=train_pipeline),
+    val=dict(pipeline=test_pipeline),
+    test=dict(pipeline=test_pipeline))
```
Comment (xvjiarui) [inline configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py:45]: These are already included in `../_base_/cityscapes.py`. We just need to modify `samples_per_gpu` and `workers_per_gpu` here.

File: configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+lr_config = dict(
+    policy='poly',
+    power=0.9,
+    by_epoch=False,
+)
+# runtime settings
+# total_epochs = 1000
+total_iters = 80000
+evaluation = dict(interval=2000, metric='mIoU')
+checkpoint_config = dict(interval=2000)
```
Comment (xvjiarui) [inline configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py:60]: We may use `_base_/schedule_80k.py` instead. 
Note that we need to set `lr=0.12`.

File: configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+# log config: log by iter.
+log_config = dict(interval=50, hooks=[dict(type='TextLoggerHook', by_epoch=False)])
+
```
Comment (xvjiarui) [inline configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py:64]: Already included in `_base_/default_runtime.py`.

File: mmseg/models/backbones/fast_scnn.py [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+from torch.nn.modules.batchnorm import _BatchNorm
+
+from mmseg.models.backbones.mobile_net_v2 import InvertedResidual
+from mmseg.models.decode_heads.psp_head import PPM
+from mmseg.ops import DepthwiseSeparableConvModule, resize
+from ..builder import BACKBONES
+
+
+class LearningToDownsample(nn.Module):
+    """Learning to downsample module."""
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:13]: We may use standard doc string here.

File: mmseg/models/backbones/fast_scnn.py [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+
+    def forward(self, x):
+        x = self.conv(x)
+        x = self.dsconv1(x)
+        x = self.dsconv2(x)
+        return x
+
+
+class GlobalFeatureExtractor(nn.Module):
+    """Global feature extractor module."""
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:56]: We may use standard doc string here.

**Comment by johnzja** [comment]
These lines are necessary under Windows (without gcc but with cl.exe).





------------------&nbsp;Original&nbsp;------------------

**Comment by johnzja** [comment]
This may not be realizable because "data" needs the definitions of "train_pipeline" and "test_pipeline". &nbsp;





------------------&nbsp;Original&nbsp;------------------

File: mmseg/ops/separable_conv_module.py (hunk 1 above)
Comment (johnzja) [reply mmseg/ops/separable_conv_module.py]: Unchanged now.

File: mmseg/models/decode_heads/sep_fcn_head.py (hunk 2 above)
Comment (johnzja) [reply mmseg/models/decode_heads/sep_fcn_head.py]: Renamed.

File: mmseg/utils/collect_env.py (hunk 3 above)
Comment (johnzja) [reply mmseg/utils/collect_env.py:50]: Necessary under Windows, but not under Linux.

File: mmseg/models/backbones/fast_scnn.py (hunk 4 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py]: Improved.

File: mmseg/models/backbones/fast_scnn.py (hunk 5 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py]: Improved.

File: mmseg/models/backbones/mobile_net_v2.py (hunk 6 above)
Comment (johnzja) [reply mmseg/models/backbones/mobile_net_v2.py]: Only the MobileNetV2 module is deleted. InvertedResidual module is directly referenced in fast-scnn.

File: configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py (hunk 7 above)
Comment (johnzja) [reply configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py]: Moved.

File: mmseg/models/backbones/fast_scnn.py (hunk 11 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py]: Improved.

File: mmseg/models/backbones/fast_scnn.py (hunk 12 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py]: Improved.

File: configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py (hunk 10 above)
Comment (johnzja) [reply configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py]: Improved.

File: configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py (hunk 9 above)
Comment (johnzja) [reply configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py]: OK

File: mmseg/utils/collect_env.py (hunk 3 above)
Comment (xvjiarui) [reply mmseg/utils/collect_env.py:49]: These could be in another PR. 
We may omit these in this PR.

File: mmseg/models/backbones/fast_scnn.py [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+        num_blocks (tuple): Tuple of ints. Each int specifies the number of
+        times each Inverted Residual module is repeated.
+
+        pool_scales (tuple): Tuple of ints. Each int specifies the parameter
+            required in 'global average pooling' within PPM.
+
+        conv_cfg (dict|None): Config of conv layers.
+        norm_cfg (dict|None): Config of norm layers.
+        act_cfg (dict): Config of activation layers.
+        align_corners (bool): align_corners argument of F.interpolate.
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:96]: No newline is needed between arguments. 
Default value is missing.

File: mmseg/models/backbones/fast_scnn.py [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+            module.
+
+        num_blocks (tuple): Tuple of ints. Each int specifies the number of
+        times each Inverted Residual module is repeated.
+
+        pool_scales (tuple): Tuple of ints. Each int specifies the parameter
+            required in 'global average pooling' within PPM.
+
+        conv_cfg (dict|None): Config of conv layers.
+        norm_cfg (dict|None): Config of norm layers.
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:94]:
```suggestion
        conv_cfg (dict | None): Config of conv layers. Default: None. 
        norm_cfg (dict  |None): Config of norm layers. Default: dict(type='BN'). 
```

File: mmseg/models/backbones/fast_scnn.py [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+            align_corners=align_corners)
+        self.out = ConvModule(
+            block_channels[2] * 2,
+            out_channels,
+            1,
+            conv_cfg=self.conv_cfg,
+            norm_cfg=self.norm_cfg,
+            act_cfg=self.act_cfg)
+
+    def _make_layer(self, inplanes, planes, blocks, t=6, stride=1):
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:139]: Docstring

File: mmseg/models/backbones/fast_scnn.py [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+                 in_channels=64,
+                 block_channels=(64, 96, 128),
+                 out_channels=128,
+                 t=6,
+                 num_blocks=(3, 3, 3),
+                 pool_scales=(1, 2, 3, 6),
+                 conv_cfg=None,
+                 norm_cfg=dict(type='BN'),
+                 act_cfg=dict(type='ReLU'),
+                 align_corners=True):
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:109]:
```suggestion
                 align_corners=False):
```

File: mmseg/models/backbones/fast_scnn.py [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+
+    def __init__(self,
+                 higher_in_channels,
+                 lower_in_channels,
+                 out_channels,
+                 scale_factor,
+                 conv_cfg=None,
+                 norm_cfg=dict(type='BN'),
+                 act_cfg=dict(type='ReLU'),
+                 align_corners=True):
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:187]:
```suggestion
                 align_corners=False):
```

File: mmseg/models/backbones/fast_scnn.py [hunk 18]
```diff
@@ -0,0 +1,361 @@
...
+            resolution branch in FFM.
+            Equal to global_in_channels.
+
+        lower_in_channels (int): Number of input channels of  the lower
+            resolution branch in FFM.
+            Equal to global_out_channels.
+
+        fusion_out_channels (int): Number of output channels of FFM.
+
+        scale_factor (int): The upsampling factor of the higher resolution
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:263]: No newline is needed between arguments.
Default value is missing.

File: mmseg/models/decode_heads/sep_fcn_head.py [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+
+        in_index(int): Correspond with 'out_indices' in FastSCNN backbone.
+
+        norm_cfg (dict|None): Config of norm layers.
+
+        align_corners (bool): align_corners argument of F.interpolate.
+
+        loss_decode(dict): Config of loss type and some
+            relevant additional options.
+    """
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/sep_fcn_head.py:31]: No newline is needed between arguments.
Default value is missing.

File: mmseg/models/backbones/mobile_net_v2.py [hunk 20]
```diff
@@ -0,0 +1,57 @@
+from mmcv.cnn import ConvModule, build_norm_layer
+from torch import nn
+
+
+class InvertedResidual(nn.Module):
```
Comment (xvjiarui) [inline mmseg/models/backbones/mobile_net_v2.py:5]: Add unit test.

File: mmseg/models/decode_heads/sep_fcn_head.py [hunk 21]
```diff
@@ -0,0 +1,52 @@
+from mmseg.ops import DepthwiseSeparableConvModule
+from ..builder import HEADS
+from .fcn_head import FCNHead
+
+
+@HEADS.register_module()
+class DepthwiseSeparableFCNHead(FCNHead):
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/sep_fcn_head.py:7]: Add unit test.

File: tests/test_models/test_backbone.py [hunk 22]
```diff
@@ -664,3 +664,25 @@ def test_resnext_backbone():
...
+    with pytest.raises(AssertionError):
+        # Fast-SCNN channel constraints.
+        FastSCNN(3, 32, 48, 64, (64, 96, 128), 127, 64, 128)
+
+    # Test FastSCNN Standard Forward
+    model = FastSCNN()
+    model.init_weights()
+    model.train()
+    num_batch_picts = 4
+    imgs = torch.randn(num_batch_picts, 3, 1024, 2048)
```
Comment (xvjiarui) [inline tests/test_models/test_backbone.py:679]: We may use smaller size for unit test.

File: tests/test_models/test_backbone.py [hunk 23]
```diff
@@ -664,3 +664,25 @@ def test_resnext_backbone():
...
+def test_fastscnn_backbone():
+    with pytest.raises(AssertionError):
+        # Fast-SCNN channel constraints.
+        FastSCNN(3, 32, 48, 64, (64, 96, 128), 127, 64, 128)
+
+    # Test FastSCNN Standard Forward
+    model = FastSCNN()
+    model.init_weights()
+    model.train()
+    num_batch_picts = 4
```
Comment (xvjiarui) [inline tests/test_models/test_backbone.py:678]:
```suggestion
    batch_size = 4
```

File: mmseg/models/backbones/fast_scnn.py [hunk 24]
```diff
@@ -0,0 +1,361 @@
...
+            dwconv layer.
+
+        out_channels (int): Number of output channels of the whole
+            'learning to downsample' module.
+
+
+        conv_cfg (dict|None): Config of conv layers.
+        norm_cfg (dict|None): Config of norm layers.
+        act_cfg (dict): Config of activation layers.
+    """
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:30]: No newline is needed between arguments.
Default value is missing.

File: mmseg/models/backbones/fast_scnn.py [hunk 25]
```diff
@@ -0,0 +1,361 @@
...
+class LearningToDownsample(nn.Module):
+    """Learning to downsample module.
+    Args:
+        in_channels (int): Number of input channels.
+
+        dw_channels1 (int): Number of output channels of the first
+            depthwise conv (dwconv) layer.
+
+        dw_channels2 (int): Number of output channels of the second
+            dwconv layer.
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:21]: We may make these two variables into a tuple.

File: mmseg/models/backbones/mobile_net_v2.py [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+
+    def __init__(self,
+                 inp,
```
Comment (xvjiarui) [inline mmseg/models/backbones/mobile_net_v2.py:8]: We may consider writing as mmcls [here](https://github.com/open-mmlab/mmclassification/blob/master/mmcls/models/backbones/mobilenet_v2.py#L14).

File: mmseg/models/backbones/fast_scnn.py (hunk 15 above)
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:139]: We may avoid using `t` here.

**Comment by xvjiarui** [comment]
Please sign CLA.

File: mmseg/utils/collect_env.py (hunk 3 above)
Comment (johnzja) [reply mmseg/utils/collect_env.py:49]: OK.

File: configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py (hunk 8 above)
Comment (johnzja) [reply configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py]: OK.

File: mmseg/models/backbones/fast_scnn.py (hunk 18 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py:262]: Fixed.

File: mmseg/models/backbones/fast_scnn.py (hunk 17 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py:186]: Fixed.

File: mmseg/models/backbones/fast_scnn.py (hunk 16 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py:108]: Fixed.

File: mmseg/models/backbones/fast_scnn.py (hunk 14 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py:93]: Fixed.

File: mmseg/models/backbones/fast_scnn.py (hunk 13 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py:95]: Fixed.

File: mmseg/models/decode_heads/sep_fcn_head.py (hunk 19 above)
Comment (johnzja) [reply mmseg/models/decode_heads/sep_fcn_head.py:31]: New lines fixed.
There is no default value specified in the code. Is it necessary to modify the code?

File: tests/test_models/test_backbone.py (hunk 23 above)
Comment (johnzja) [reply tests/test_models/test_backbone.py:678]: Fixed.

File: mmseg/models/backbones/mobile_net_v2.py (hunk 20 above)
Comment (johnzja) [reply mmseg/models/backbones/mobile_net_v2.py]: Fixed.

File: mmseg/models/backbones/fast_scnn.py (hunk 24 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py]: Fixed.

File: tests/test_models/test_backbone.py (hunk 22 above)
Comment (johnzja) [reply tests/test_models/test_backbone.py]: Fixed.

File: mmseg/models/decode_heads/sep_fcn_head.py (hunk 21 above)
Comment (johnzja) [reply mmseg/models/decode_heads/sep_fcn_head.py:7]: Fixed.

File: mmseg/models/backbones/fast_scnn.py (hunk 25 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py]: Fixed.

File: mmseg/models/backbones/fast_scnn.py [hunk 27]
```diff
@@ -0,0 +1,355 @@
...
+from mmseg.ops import DepthwiseSeparableConvModule, InvertedResidual, resize
+from ..builder import BACKBONES
+
+
+class LearningToDownsample(nn.Module):
+    """Learning to downsample module.
+
+    Args:
+        in_channels (int): Number of input channels.
+        dw_channels (tuple): Number of output channels of the first and
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:16]:
```suggestion
        dw_channels (tuple[int]): Number of output channels of the first and
```

File: mmseg/models/backbones/fast_scnn.py [hunk 28]
```diff
@@ -0,0 +1,355 @@
...
+        in_channels (int): Number of input channels of the GFE module.
+        block_channels (tuple): Tuple of ints. Each int specifies the
+            number of output channels of each Inverted Residual module.
+        out_channels(int): Number of output channels of the GFE module.
+        expand_ratio (int): upsampling factor of each Inverted Residual
+            module.
+        num_blocks (tuple): Tuple of ints. Each int specifies the number of
+        times each Inverted Residual module is repeated.
+        pool_scales (tuple): Tuple of ints. Each int specifies the parameter
+            required in 'global average pooling' within PPM.
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:84]: Default values are missing.

File: mmseg/models/decode_heads/sep_fcn_head.py [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+            relevant additional options.
+    """
+
+    def __init__(self, **kwargs):
+        super(DepthwiseSeparableFCNHead, self).__init__(**kwargs)
+        self.convs[0] = DepthwiseSeparableConvModule(
+            self.in_channels,
+            self.channels,
+            kernel_size=3,
+            padding=1,
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/sep_fcn_head.py:34]:
```suggestion
            kernel_size=self.kernel_size,
            padding=self.kernel_size//2,
```

File: mmseg/models/decode_heads/sep_fcn_head.py [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+            norm_cfg=self.norm_cfg)
+        for i in range(1, self.num_convs):
+            self.convs[i] = DepthwiseSeparableConvModule(
+                self.channels,
+                self.channels,
+                kernel_size=3,
+                padding=1,
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/sep_fcn_head.py:41]:
```suggestion
                kernel_size=self.kernel_size,
                padding=self.kernel_size//2,
```

File: mmseg/ops/inverted_residual_module.py [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+
+    def __init__(self,
+                 inp,
+                 oup,
+                 stride,
+                 expand_ratio,
+                 dilation=1,
+                 conv_cfg=None,
+                 norm_cfg=dict(type='BN'),
+                 act_cfg=dict(type='ReLU6')):
```
Comment (xvjiarui) [inline mmseg/ops/inverted_residual_module.py:28]: We may use similar variable name as [mmcls](https://github.com/open-mmlab/mmclassification/blob/master/mmcls/models/backbones/mobilenet_v2.py#L14).

File: mmseg/models/backbones/fast_scnn.py [hunk 32]
```diff
@@ -0,0 +1,355 @@
...
+        super(FastSCNN, self).__init__()
+        if global_in_channels != higher_in_channels:
+            raise AssertionError('Global Input Channels must be the same \
+                                 with Higher Input Channels!')
+        elif global_out_channels != lower_in_channels:
+            raise AssertionError('Global Output Channels must be the same \
+                                with Lower Input Channels!')
+        if scale_factor != 4:
+            raise AssertionError('Scale-factor must compensate the \
+                            downsampling factor in the GFE module!')
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:298]: Why only 4?

File: configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py [hunk 33]
```diff
@@ -0,0 +1,11 @@
+_base_ = [
+    '../_base_/models/fast_scnn.py', '../_base_/datasets/cityscapes.py',
+    '../_base_/default_runtime.py', '../_base_/schedules/schedule_80k.py'
+]
+cudnn_benchmark = True
```
Comment (xvjiarui) [inline configs/fastscnn/fast_scnn_4x8_80k_lr0.12_cityscapes.py:5]: This could be omitted.

File: mmseg/models/backbones/fast_scnn.py [hunk 34]
```diff
@@ -0,0 +1,355 @@
...
+        lower_res_feature = self.conv_lower_res(lower_res_feature)
+
+        higher_res_feature = self.conv_higher_res(higher_res_feature)
+        out = higher_res_feature + lower_res_feature
+        return self.relu(out)
+
+
+@BACKBONES.register_module()
+class FastSCNN(nn.Module):
+    """Fast-SCNN Backbone.
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:229]: Missing blank line.

File: mmseg/models/decode_heads/sep_fcn_head.py [hunk 35]
```diff
@@ -0,0 +1,50 @@
...
+    Args:
+        in_channels(int): Number of output channels of FFM.
+        channels(int): Number of middle-stage channels in the decode head.
+        concat_input(bool): Whether to concatenate original decode input into
+            the result of several consecutive convolution layers.
+            Default: True.
+        num_classes(int): Used to determine the dimension of
+            final prediction tensor.
+        in_index(int): Correspond with 'out_indices' in FastSCNN backbone.
+        norm_cfg (dict|None): Config of norm layers.
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/sep_fcn_head.py:21]:
```suggestion
        norm_cfg (dict | None): Config of norm layers.
```

File: mmseg/ops/inverted_residual_module.py [hunk 36]
```diff
@@ -0,0 +1,70 @@
+from mmcv.cnn import ConvModule, build_norm_layer
+from torch import nn
+
+
+class InvertedResidual(nn.Module):
+    """Inverted Residual Module
+    Args:
+        inp (int): input channels.
```
Comment (xvjiarui) [inline mmseg/ops/inverted_residual_module.py:8]: Capitalize first letter.

File: mmseg/models/backbones/fast_scnn.py (hunk 34 above)
Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py:234]: Added.

File: mmseg/models/backbones/fast_scnn.py [hunk 37]
```diff
@@ -0,0 +1,361 @@
...
+    def _make_layer(self, inplanes, planes, blocks, expand_ratio=6, stride=1):
+        layers = [
+            InvertedResidual(
+                inplanes, planes, stride, expand_ratio, norm_cfg=self.norm_cfg)
+        ]
+        for i in range(1, blocks):
+            layers.append(
+                InvertedResidual(
+                    planes, planes, 1, expand_ratio, norm_cfg=self.norm_cfg))
+        return nn.Sequential(*layers)
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:148]: Variable names should be consistent with `InvertedResidual`.

Comment (johnzja) [reply mmseg/models/backbones/fast_scnn.py]: OK.

File: mmseg/models/backbones/fast_scnn.py [hunk 38]
```diff
@@ -0,0 +1,374 @@
...
+
+    Args:
+        in_channels (int): Number of input channels of the GFE module.
+            Default: 64
+        block_channels (tuple[int]): Tuple of ints. Each int specifies the
+            number of output channels of each Inverted Residual module.
+            Default: (64, 96, 128)
+        out_channels(int): Number of output channels of the GFE module.
+            Default: 128
+        expand_ratio (int): Upsampling factor of each Inverted Residual
```
Comment (xvjiarui) [inline mmseg/models/backbones/fast_scnn.py:82]: Upsampling is not correct.

File: configs/fastscnn/README.md [hunk 39]
```diff
@@ -0,0 +1,27 @@
...
+  volume    = {abs/1902.04502},
+  year      = {2019},
+  url       = {http://arxiv.org/abs/1902.04502},
+  archivePrefix = {arXiv},
+  eprint    = {1902.04502},
+  timestamp = {Tue, 21 May 2019 18:03:38 +0200},
+  biburl    = {https://dblp.org/rec/journals/corr/abs-1902-04502.bib},
+  bibsource = {dblp computer science bibliography, https://dblp.org}
+}
+```
```
Comment (hellock) [inline configs/fastscnn/README.md:20]:
```
@article{poudel2019fast,
  title={Fast-scnn: Fast semantic segmentation network},
  author={Poudel, Rudra PK and Liwicki, Stephan and Cipolla, Roberto},
  journal={arXiv preprint arXiv:1902.04502},
  year={2019}
}
```

**Comment by johnzja** [comment]
Can we merge now?

File: mmseg/ops/inverted_residual_module.py [hunk 40]
```diff
@@ -0,0 +1,73 @@
+from mmcv.cnn import ConvModule, build_norm_layer
+from torch import nn
+
+
+class InvertedResidual(nn.Module):
```
Comment (hellock) [inline mmseg/ops/inverted_residual_module.py:5]: This class is not an op but a common block. It can be moved to `models.utils`.

Comment (johnzja) [reply mmseg/ops/inverted_residual_module.py]: "Inverted Residual" has been moved to models.utils.

**Comment by hellock** [comment]
There are conflicts with the master branch.

**Comment by johnzja** [comment]
Now the master branch is correct.

File: tests/test_ops/test_inverted_residual_module.py [hunk 41]
```diff
@@ -0,0 +1,40 @@
+import pytest
```
Comment (hellock) [inline tests/test_ops/test_inverted_residual_module.py:1]: Since InvertedResidual is moved to utils, the test case should also be moved accordingly.

Comment (johnzja) [reply tests/test_ops/test_inverted_residual_module.py]: moved.

**Comment by johnzja** [comment]
Can we merge now?

**Comment by OpenMMLab-Assistant-007** [comment]
Hi!
@johnzja 
First of all, we want to express our gratitude for your significant PR in the OpenMMLab project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤

**Pull request merged**